### ISSUE TITLE
anomalous result

### ISSUE THREAD

**Issue by progman1** [srenevey/neuro#4]
I've been using example/xor.rs as a regression test and hit a problem that went away while
trying to track it down.
but It can be demonstrated with a slightly altered version of xor.rs:
the model setup/creation is repeated thereby abandoning the first model before it is used:
```
use neuro::activations::Activation;
use neuro::data::{TabularDataSet, DataSet};
use neuro::errors::*;
use neuro::initializers::*;
use neuro::layers::Dense;
use neuro::losses;
use neuro::metrics;
use neuro::models;
use neuro::optimizers::SGD;
use neuro::tensor::*;

fn main() -> Result<(), Error> {
    // Create the dataset
    let input_values = [0., 0., 0., 1., 1., 0., 1., 1.];
    let x_train = Tensor::new(&input_values, Dim::new(&[2, 1, 1, 4]));
    let output_values = [0., 1., 1., 0.];
    let y_train = Tensor::new(&output_values, Dim::new(&[1, 1, 1, 4]));
    let data = TabularDataSet::from_tensor(x_train.copy(), y_train.copy(), None, None, None, None)?;

    // Create the neural network and add two layers
    let mut nn = models::Network::new(Dim::new(&[2, 1, 1, 1]), losses::BinaryCrossEntropy, SGD::new(0.1), None)?;
    nn.add(Dense::with_param(2, Activation::ReLU, Initializer::Constant(0.01), Initializer::Zeros));
//here! including the following line triggers the misbehaviour
    nn.add(Dense::new(1, Activation::Sigmoid));

    let input_values = [0., 0., 0., 1., 1., 0., 1., 1.];
    let x_train = Tensor::new(&input_values, Dim::new(&[2, 1, 1, 4]));
    let output_values = [0., 1., 1., 0.];
    let y_train = Tensor::new(&output_values, Dim::new(&[1, 1, 1, 4]));
    let data = TabularDataSet::from_tensor(x_train.copy(), y_train.copy(), None, None, None, None)?;

    // Create the neural network and add two layers
    let mut nn = models::Network::new(Dim::new(&[2, 1, 1, 1]), losses::BinaryCrossEntropy, SGD::new(0.1), None)?;
    nn.add(Dense::with_param(2, Activation::ReLU, Initializer::Constant(0.01), Initializer::Zeros));
    nn.add(Dense::new(1, Activation::Sigmoid));

    // Fit the model
    nn.fit(&data, 4, 2000, Some(200), Some(vec![metrics::Metrics::Accuracy]));

    // Compute the output for the training data
    let predictions = nn.predict(&x_train);
    println!("Predictions:");
    Tensor::print_tensor(&predictions);
    Ok(())
}
```
the results are different to the output of the original xor.rs.

**Comment by progman1** [comment]
incidentally I found that 
```
AF_OPENCL_DEVICE_TYPE=CPU time cargo run --example xor --release
```
with original xor.rs, also produced strange results - it ran almost instantaneously and gave non-standard output.
perhaps an ArrayFire bug?

**Comment by srenevey** [comment]
For the first comment: there seems to be an issue with the random number generation used to initialize the weights and biases. A quick dirty fix is to add `set_seed(0)` in the model `new` method. Using a random seed seems not to work. I will investigate this further.

For the second comment: I have the same issue on my side. It might be that you need to install the OpenCL runtime as mentioned in the [ArrayFire doc](http://arrayfire.org/docs/installing.htm):

> Make sure you have the latest device drivers installed on your system before using ArrayFire. If you are going to be targeting the CPU using ArrayFire’s OpenCL backend, you will need to have the OpenCL runtime installed on your system. Drivers and runtimes should be downloaded and installed from your device vendor’s website.

What OS do you use? Have you by any chance tested the crate with CUDA?

**Comment by srenevey** [comment]
The issue for the xor example is fixed in 2c6d04a115f6e50b541a8905bcb0be3c8ffdad79. The way the weights of the network were initialized was problematic due to the small number of parameters, resulting in vanishing/exploding gradient scenarios. Training a XOR gate is not as simple as it looks and tuning the weight initialization, optimizer, and number of training epochs might be necessary for better results.

**Comment by progman1** [comment]
> What OS do you use? Have you by any chance tested the crate with CUDA?

I use x86_64-linux-gnu and only AMD/openCL.
I have libafcpu.so installed so I think the runtime switching between
implementations is there.
but you have that library also and still see the issue?

**Comment by progman1** [comment]
in regards to your fix of xor.rs I am confused - the problem stemmed from some
hidden state lurking inside the code, no?
and the quick fix for that was using a random seed of 0?
the code change I made highlighted the running of code whose values
are abandoned yet which
turn out to have an underlying effect on the next values produced,
which suggests shared state.

**Comment by srenevey** [comment]
> but you have that library also and still see the issue?

Yes, but I've found a way to make it work: try adding `use arrayfire::{set_backend, Backend};` at the beginning of xor.rs and `set_backend(Backend::CPU)` at the beginning of the main function.

> in regards to your fix of xor.rs I am confused - the problem stemmed from some
hidden state lurking inside the code, no?
and the quick fix for that was using a random seed of 0?

In the previous version, the weights of the two units in the hidden layer were initialized with a constant value of 0.01. The weight of the output layer was initialized randomly using a HeNormal initializer. That is, it was drawn from a normal distribution with mean 0 and standard deviation sqrt(2/fan-in) = 1. Now, using a fixed seed with with a pseudo-random generator results in random numbers generated in the same sequence every time the program is run. For a seed of 42 for instance, the sequence will be, say, 0.1 -> -1.3 -> -0.12 ... every time. Hence, when the neural network was created twice, the initial weight of the output layer of the first network was 0.1 and that of the second network -1.3. Turns out that the network converged for 0.1 but not for -1.3. Using a seed of 0 resulted, by chance, in the two networks to converge. 
A proper fix is to tune the initilizer to obtain initial weights which are more likely to result in convergence. Also, in order to avoid having the same sequence of random numbers every time the program is run, a random seed is used (models.rs has been updated). This will help spotting that kind of bugs in the future.

Does that help?

**Comment by progman1** [comment]
yes indeed. thanks for taking the time to explain that one.

**Comment by progman1** [comment]
Perhaps it might apply to only myself and my library wrapper but I
used the original xor.rs
as a regression test - the repeatable results were invaluable in
assuring no errors slipped in ( and they often did) unnoticed.
so I would encourage introduction of a means of running with
pseudo-randomness (perhaps an environment variable) over a default of
true.
